src/core/algorithms/dd/split/split.cpp

```diff
@@ -299,96 +353,75 @@ double Split::CalculateDistance(model::ColumnIndex column_index,
 }
 
 // must be inline for optimization (gcc 11.4.0)
-inline bool Split::CheckDF(DF const& dif_func, std::pair<std::size_t, std::size_t> tuple_pair) {
-    for (model::ColumnIndex column_index = 0; column_index < num_columns_; column_index++) {
-        double const dif = distances_[column_index][tuple_pair.first][tuple_pair.second];
+inline bool Split::CheckDFConstraint(DFConstraint const& dif_constraint,
+                                     model::ColumnIndex column_index,
+                                     std::pair<std::size_t, std::size_t> tuple_pair) {
+    std::vector<ClusterIndex> const& cur_index = plis_[column_index].GetInvertedIndex();
+    ClusterIndex const first_cluster = cur_index[tuple_pair.first];
+    ClusterIndex const second_cluster = cur_index[tuple_pair.second];
+    ClusterIndex const min_cluster = std::min(first_cluster, second_cluster);
+    ClusterIndex const max_cluster = std::max(first_cluster, second_cluster);
+    double const dif = distances_[column_index][min_cluster][max_cluster - min_cluster];
 
-        if (type_ids_[column_index] == +model::TypeId::kDouble) {
-            if (!dif_func[column_index].Contains(dif)) {
-                return false;
-            }
-        } else {
-            if (dif < dif_func[column_index].lower_bound ||
-                dif > dif_func[column_index].upper_bound) {
-                return false;
-            }
-        }
+    if (type_ids_[column_index] == +model::TypeId::kDouble) {
+        return dif_constraint.Contains(dif);
     }
-    return true;
+
+    return dif >= dif_constraint.lower_bound && dif <= dif_constraint.upper_bound;
 }
 
-bool Split::VerifyDD(DF const& lhs, DF const& rhs) {
-    for (std::size_t i = 0; i < num_rows_; i++) {
-        for (std::size_t j = i + 1; j < num_rows_; j++) {
-            if (CheckDF(lhs, {i, j}) && !CheckDF(rhs, {i, j})) return false;
+// must be inline for optimization (gcc 11.4.0)
+inline bool Split::CheckDF(DF const& dif_func, std::pair<std::size_t, std::size_t> tuple_pair) {
+    for (model::ColumnIndex column_index = 0; column_index < num_columns_; column_index++) {
+        if (!CheckDFConstraint(dif_func[column_index], column_index, tuple_pair)) {
+            return false;
         }
     }
     return true;
 }
```

Comment:
```cpp
    return std::ranges::all_of(
        std::views::iota(0, num_columns_),
        [&](model::ColumnIndex column_index) {
            return CheckDFConstraint(dif_func[column_index], column_index, tuple_pair);
        });
```

Reply:
check performance without inline with std::ranges algo btw, maybe this time compiler will inline it